feat(quantic): new feedback modal for RGA

erocheleau opened the pull request:
Mockups: https://www.figma.com/design/l63i3ZXWrVxldOR4WUBKS1/CRGA-Component?node-id=5514-14647&t=pOZ5lDKywVYJCqgI-0

SFINT-5555:

New Quantic RGA Feedback modal

![image](https://github.com/user-attachments/assets/4ffd312c-e0ce-475d-87f8-17cbdb288fc7)

The top 4 questions are required, and you have an example of how we tell the user when they are required and missing.

The bottom two text fields are optional.

When you press send feedback there will be the same screen as before thanking you for your feedback.

If you press skip it closes the modal. And nothing you've selected is "remembered" if you re-open it.

The modal now opens on both like and dislike (thumbs up and thumbs down), and as part of the event we send, we specify if the feedback was positive or negative. We still send a separate event already when you press thumbs up or down.

You can "change your mind", even if you send feedback, you can still change your thumbs up to a thumbs down. but it will NOT re-open the modal if you've already submitted feedback once. 
Also if you click a 2nd time on thumbs down for example we won't re-open the modal.

Everything is tested through unit tests since this is only the "content" of the modal, and not the feedback process itself that we're testing.

mmitiche commented:
> I changed it so that you can "change your mind", even if you send feedback, you can still change your thumbs up to a thumbs down. and it will re-open the modal. But if you click a 2nd time on thumbs down for example we won't re-open the modal.

Is this the same behaviour we have in Atomic?

SimonMilord commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -1442,4 +1463,74 @@
+  </labels>
+  <labels>
+    <fullName>quantic_FeedbackQuestionHallucination</fullName>
+    <value>Is the answer free of hallucinated content?</value>
```
```suggestion
    <value>Is the answer free of hallucinated content?</value>
```

Is this not too "technical" for users? Seems like a weird way to just say something like: "Is the answer accurate?"
Hallucinations are LLM/Generative AI jargon IMO so I would be tempted to keep it simple. What do you think?

erocheleau replied:
Nope the questions have been decided by ML. The target audience is a lot more "Subject matter experts" than general users AFAIK, even if we present it to everyone.
Accurate is already pretty much covered by the other questions too.

SimonMilord commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -1442,4 +1463,74 @@
+  </labels>
+  <labels>
+    <fullName>quantic_FeedbackQuestionDocumented</fullName>
+    <value>Can the question be answered by the documentation?</value>
```
What is the purpose of this question? 😂  I feel like most users who will ask questions to the model might not check the doc X)

erocheleau replied:
It's for subject matter experts to tell us.

SimonMilord commented on packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:
```diff
@@ -1442,4 +1463,74 @@
+  </labels>
+  <labels>
+    <fullName>quantic_FeedbackDocumentUrl</fullName>
+    <value>Link to correct answer</value>
```
```suggestion
    <value>Link to the correct answer</value>
```

erocheleau replied:
That's coming from the mockup, not my decision.

SimonMilord commented on packages/quantic/force-app/main/translations/fr.translation-meta.xml:
```diff
@@ -796,4 +808,36 @@
+    <name>quantic_FeedbackQuestionTopic</name>
+  </customLabels>
+  <customLabels>
+    <label>La réponse est-elle exempte de contenu halluciné?</label>
```
same comment here as above in the english file

erocheleau replied:
Same answer :)

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/quanticGeneratedAnswer.js:
```diff
@@ -297,16 +315,16 @@ export default class QuanticGeneratedAnswer extends LightningElement {
       // @ts-ignore
-      await FeedbackModal.open({
-        label: this.labels.feedback,
+      await FeedbackModalQna.open({
```
if we are no longer opening the feedbackModal, but the FeedbackModalQna here shouldn't we remove the options getter and associated labels?

erocheleau replied:
wut?

erocheleau commented:
> > I changed it so that you can "change your mind", even if you send feedback, you can still change your thumbs up to a thumbs down. and it will re-open the modal. But if you click a 2nd time on thumbs down for example we won't re-open the modal.
> 
> Is this the same behaviour we have in Atomic?

Yea you were right to ask, I updated the PR.

If you've submitted the feedback once, thumbs up or down, you won't see the modal again until after you do a new query.

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/quanticFeedbackModalQna.js:
```diff
@@ -0,0 +1,192 @@
+    additionalNotesExplanation,
+  };
+
+  /**
+   * @api
+   * The function that will be executed when the feedback is submitted.
+   * @type {function}
+   */
+  @api handleSubmit;
+
+  /**
+   * @api
+   * If the feedback modal was opened on the thumbs up. (otherwise it was on a thumbs down).
+   * @type {boolean}
+   */
+  @api liked;
```
```suggestion
  /**
   * @api
   * The function that will be executed when the feedback is submitted.
   * @type {function}
   */
  @api handleSubmit;
  /**
   * @api
   * If the feedback modal was opened on the thumbs up. (otherwise it was on a thumbs down).
   * @type {boolean}
   */
  @api liked;
```

SimonMilord replied:
```suggestion
  /**
   * @api
   * The function that will be executed when the feedback is submitted.
   * @type {function}
   */
  @api handleSubmit;
  /**
   * @api
   * If the feedback modal was opened on the thumbs up. (otherwise it was on a thumbs down).
   * @type {boolean}
   */
  @api liked;
```

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/quanticFeedbackModalQna.js:
```diff
@@ -0,0 +1,192 @@
+    this.isFeedbackSubmitted = true;
+  }
+
+  /**
+   * Returns the right HTML template.
+   */
```
```suggestion
```

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/quanticFeedbackModalQna.js:
```diff
@@ -0,0 +1,192 @@
+    let valid = true;
+    const feedbackQuestions = this.feedbackQuestionsRadios;
+    feedbackQuestions.forEach((question) => {
+      if (question?.checkValidity() !== true) {
```
where is `checkValidity()` defined? Did I miss something?

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/quanticFeedbackModalQna.js:
```diff
@@ -0,0 +1,192 @@
+    return this.template.querySelectorAll('lightning-radio-group');
+  }
+
+  get feedbackFormIsValid() {
```
nice!

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/__tests__/quanticFeedbackModalQna.test.js:
```diff
@@ -0,0 +1,299 @@
+  skipButton: 'lightning-button.skip',
+  submitButton: 'lightning-button.submit-feedback',
+  requiredMessages: 'slds-form-element__help',
+  successMesage: '.feedback-modal-body__success-message',
```
```suggestion
  successMessage: '.feedback-modal-body__success-message',
```

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/__tests__/quanticFeedbackModalQna.test.js:
```diff
@@ -0,0 +1,299 @@
+  additionalDetailsTextarea: 'lightning-textarea[data-name="details"]',
+  skipButton: 'lightning-button.skip',
+  submitButton: 'lightning-button.submit-feedback',
+  requiredMessages: 'slds-form-element__help',
```
this selector doesn't seem used anywhere, is it extra or are you missing something in a test?

erocheleau replied:
yea no you're correct, it was before I realized the inputs were just stubs in tests (so even if they are invalid they don't display their required message) I was trying to make a test to make sure they were displayed, I'll remove it.

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/__tests__/quanticFeedbackModalQna.test.js:
```diff
@@ -0,0 +1,299 @@
+    );
+    expect(answerEvaluationQuestions.length).toBe(4);
+
+    const questionTopicRadio = element.shadowRoot.querySelector(
+      selectors.questionTopicRadio
+    );
+    expect(questionTopicRadio).not.toBeNull();
+
+    const questionHallucinationRadio = element.shadowRoot.querySelector(
+      selectors.questionHallucinationRadio
+    );
+    expect(questionHallucinationRadio).not.toBeNull();
+
+    const questionDocumentedRadio = element.shadowRoot.querySelector(
+      selectors.questionDocumentedRadio
+    );
+    expect(questionDocumentedRadio).not.toBeNull();
+
+    const questionReadableRadio = element.shadowRoot.querySelector(
+      selectors.questionReadableRadio
+    );
+    expect(questionReadableRadio).not.toBeNull();
+
+    const documentUrlInput = element.shadowRoot.querySelector(
+      selectors.documentUrlInput
+    );
+    expect(documentUrlInput).not.toBeNull();
+
+    const additionalDetailsInput = element.shadowRoot.querySelector(
+      selectors.additionalDetailsTextarea
+    );
+    expect(additionalDetailsInput).not.toBeNull();
+
+    const skipButton = element.shadowRoot.querySelector(selectors.skipButton);
+    expect(skipButton).not.toBeNull();
+
+    const submitButton = element.shadowRoot.querySelector(
+      selectors.submitButton
+    );
+    expect(submitButton).not.toBeNull();
```
I would use an array of selectors here to just check with a forEach()

erocheleau replied:
No, because the tests if an element fail to render, would fail inside the loop and we would have no idea which element didn't render correctly.

If we keep it like this, we know exactly which expect failed and which element it was.

In tests, you want to be explicit even if it means writing more code.

SimonMilord replied:
Are you sure? Im pretty sure I tried it somewhere before and despite the loop it was still logging in error the component in the array that was faulty...

did that in the `hostedSearchPageRenderer.spec.ts`

<img width="732" alt="image" src="https://github.com/user-attachments/assets/26686099-c2d0-4bf4-a4d5-0f321d066fd2">

erocheleau replied:
ah because you put the `it` in the loop, your example creates 1 test per component.

My test tests all "default" components in 1 test, I don't want to have 1 test per component, otherwise it will re-create and render the component X times, instead of loading once and checking for the components to be there.

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/__tests__/quanticFeedbackModalQna.test.js:
```diff
@@ -0,0 +1,299 @@
+
+  describe('when submitting the feedback', () => {
+    describe('with missing required answers', () => {
+      it('should not submit the feedback and not show the success screen', async () => {
```
As mentionned I would add a check for the required message here

erocheleau replied:
I can't. Jest lightning-* components don't actually "exist". They are only stubs in the DOM with some public attributes accessible to tests.

Meaning they don't actually react to display a required message while in unit tests.

SimonMilord replied:
ok

SimonMilord commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/__tests__/quanticFeedbackModalQna.test.js:
```diff
@@ -0,0 +1,299 @@
+            details: expectedDetails,
+          })
+        );
+      });
+    });
+  });
+  describe('when clicking the skip button', () => {
+    it('should not submit the feedback and not execute the handleSubmit function', async () => {
+      const element = createTestComponent();
+      await flushPromises();
```
```suggestion
      });
    });
  });
  
  describe('when clicking the skip button', () => {
    it('should not submit the feedback and not execute the handleSubmit function', async () => {
      const element = createTestComponent();
      await flushPromises();
```

mmitiche commented:
I'm noticing some differences with the mockups we have, while they maybe are not all adjustable because we are using SF components, some others I think are adjustable:

Mockups:
<img width="502" alt="Screenshot 2024-07-15 at 10 40 10 AM" src="https://github.com/user-attachments/assets/d88f539c-3752-4efe-973c-386566ffd469">

Real compomnent:
<img width="855" alt="Screenshot 2024-07-15 at 10 43 47 AM" src="https://github.com/user-attachments/assets/18505ce6-043d-47e9-a357-919fe103a6e9">

- The label `Answer Evaluation` is not with the right color, it currently a weak text color while in the mockup it's not.
- The label of the two text inputs should be bold and with a black color, it's currently not bold and with a weak text color.
- The label in the modal bottom should be `Required fields` instead of `This field is required`. not a big deal but it's better to stay consistent

mmitiche commented:
one of the other differences is that the end of the text inputs is not well aligned with the end of the radio buttons input like in the mokups, this happens in narrower screens, maybe we could find a better way to align this:

<img width="587" alt="Screenshot 2024-07-15 at 10 55 37 AM" src="https://github.com/user-attachments/assets/1414a65f-a3e6-4ab5-ad7e-8cfca770ef3a">
<img width="531" alt="Screenshot 2024-07-15 at 10 56 06 AM" src="https://github.com/user-attachments/assets/52b4466f-2c50-44c3-8b13-30a856bc36cc">


https://github.com/user-attachments/assets/3fc377d3-d2bd-48d2-b29f-4ae17c36c6bf

erocheleau commented:
> one of the other differences is that the end of the text inputs is not well aligned with the end of the radio buttons input like in the mokups, this happens in narrower screens, maybe we could find a better way to align this:
> 
> <img alt="Screenshot 2024-07-15 at 10 55 37 AM" width="587" src="https://private-user-images.githubusercontent.com/86681870/348752069-1414a65f-a3e6-4ab5-ad7e-8cfca770ef3a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwNTU5MDEsIm5iZiI6MTcyMTA1NTYwMSwicGF0aCI6Ii84NjY4MTg3MC8zNDg3NTIwNjktMTQxNGE2NWYtYTNlNi00YWI1LWFkN2UtOGNmY2E3NzBlZjNhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE1VDE1MDAwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE0MTg4MWQzNGM1NTNmYmFmMzU5NjhiNjBjODhiZjQxMTUxMTQyZDI5NTNhNjFjNjI1ODgyZTVmMDYzZjQ5MmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.g5A7oLHgRLNXErPNwSBRse0RUF4l1zBqaVTAEWnTZIY"> <img alt="Screenshot 2024-07-15 at 10 56 06 AM" width="531" src="https://private-user-images.githubusercontent.com/86681870/348752074-52b4466f-2c50-44c3-8b13-30a856bc36cc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwNTU5MDEsIm5iZiI6MTcyMTA1NTYwMSwicGF0aCI6Ii84NjY4MTg3MC8zNDg3NTIwNzQtNTJiNDQ2NmYtMmM1MC00NGMzLThiMTMtMzBhODU2YmMzNmNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE1VDE1MDAwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkMzllYzg5MmI5YjRjZjdmMzNkMGRmNmEzYTdjNTBlMjI0NDdjYzA0ZmE2M2ExODBkZDg4MTA2MTQ0MDNhMDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cYWt8Yi15jLXXDfYB_vPzRjjDOoKLqZDHfknP5gm8Uk">
>  Screen.Recording.2024-07-15.at.10.57.54.AM.mov

I'll check for narrower screens thanks. 
Didn't think anything I was using would be impacted by that.

erocheleau commented:
> I'm noticing some differences with the mockups we have, while they maybe are not all adjustable because we are using SF components, some others I think are adjustable:
> 
> Mockups: <img alt="Screenshot 2024-07-15 at 10 40 10 AM" width="502" src="https://private-user-images.githubusercontent.com/86681870/348748444-d88f539c-3752-4efe-973c-386566ffd469.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwNTU2ODAsIm5iZiI6MTcyMTA1NTM4MCwicGF0aCI6Ii84NjY4MTg3MC8zNDg3NDg0NDQtZDg4ZjUzOWMtMzc1Mi00ZWZlLTk3M2MtMzg2NTY2ZmZkNDY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE1VDE0NTYyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE2NmJiYmM4MDMyNDVlYTMyNmZlYTM3MmY0ZGMxN2E3NjBkY2IyNGZjY2ZmZTVhNGVmMzFmMGUyM2YzYjMyNjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.pOCqF_bBVEU4L3E8N5t5d1975NjvCKiMWHmBb0h04so">
> 
> Real compomnent: <img alt="Screenshot 2024-07-15 at 10 43 47 AM" width="855" src="https://private-user-images.githubusercontent.com/86681870/348748510-18505ce6-043d-47e9-a357-919fe103a6e9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwNTU2ODAsIm5iZiI6MTcyMTA1NTM4MCwicGF0aCI6Ii84NjY4MTg3MC8zNDg3NDg1MTAtMTg1MDVjZTYtMDQzZC00N2U5LWEzNTctOTE5ZmUxMDNhNmU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE1VDE0NTYyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdkODkwYTllNDE4NzliM2E3ZjEwMGFhYjQwY2VhYzE1NTQ4NDdmNmI3YzdkNjQyNDBlYTMyNjNjMTRkOTllNTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rgPTMuIdBspWOZr6PPu25mkkLnTkj__F11GRp6QYmqs">
> 
> * The label `Answer Evaluation` is not with the right color, it currently a weak text color while in the mockup it's not.
> * The label of the two text inputs should be bold and with a black color, it's currently not bold and with a weak text color.
> * The label in the modal bottom should be `Required fields` instead of `This field is required`. not a big deal but it's better to stay consistent

For the labels colors, it's expected and validated with Vyka already.

The things is the "Answer Evaluation" is a style from slds as a title/sub-title.
The other two (link to correct answer and additional notes) are input labels (a styling I cannot control.

I think this is simply due to the fact that we don't have a "Quantic" version of the mockup.

Good catch on the text for "Required fields" I'll update it.

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/__tests__/quanticFeedbackModalQna.test.js:
```diff
@@ -0,0 +1,299 @@
+    jest.clearAllMocks();
+  });
+
+  it('loads correctly', async () => {
```
It already says the name of the component in the describe above.

But I changed it to `should properly render` if you prefer that.

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/quanticFeedbackModalQna.js:
```diff
@@ -0,0 +1,188 @@
+  }
+
+  detailsChangeHandler(event) {
+    this.detailsValue = event?.target?.value;
```
same thing here.

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/quanticFeedbackModalQna.js:
```diff
@@ -0,0 +1,188 @@
+    {label: this.labels.answerNo, value: 'no'},
+  ];
+
+  /** @type {string} */
+  error;
```
I don't see that this is being used anywhere?

erocheleau replied:
yea it's in the HTML on the `onclick` of the close button or skip button

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/success.html:
usually we add the templates in a folder named `templates` for a better organization, it would be nice to keep following that pattern.

mmitiche replied:
also I don't see the success screen in the mockups, so just want to double check, do we really want it? it exists in Atomic as well?

erocheleau replied:
Confirmed yes Atomic has it.

But I will change the message on it, since it was to be changed.

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/quanticFeedbackModalQna.js:
```diff
@@ -0,0 +1,190 @@
+   * If the feedback modal was opened on the thumbs up. (otherwise it was on a thumbs down).
+   * @type {boolean}
+   */
+  @api liked;
```
I don't think that this property is really necessary for this component as I see that it is only used when calling the `handleSubmit` method which passed from the QuanticGeneratedAnswer component, and in the QuanticGeneratedAnswer we have the information about whether the answer is liked or not, so no need it to pass it here.

erocheleau replied:
yea good simplification!

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/quanticGeneratedAnswer.js:
```diff
@@ -286,9 +292,21 @@ export default class QuanticGeneratedAnswer extends LightningElement {
+      this._disliked = false;
+      this.generatedAnswer.like?.();
+      // @ts-ignore
+      await FeedbackModalQna.open({
+        size: 'small',
+        liked: true,
+        label: this.labels.feedbackHelpUsImprove,
+        handleSubmit: this.submitFeedback.bind(this),
+      });
```
```suggestion
      await FeedbackModalQna.open({
        size: 'small',
        label: this.labels.feedbackHelpUsImprove,
        handleSubmit: (payload) => {this.submitFeedback(...payload, helpful: true)},
      });
```
you may also need to transform `submitFeedback` to an arrow function to avoid the `bind(this)`

erocheleau replied:
Interesting, this implies the feedback modal doesn't know if it was triggered by like or dislike

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/quanticGeneratedAnswer.js:
```diff
@@ -297,16 +315,16 @@ export default class QuanticGeneratedAnswer extends LightningElement {
+      await FeedbackModalQna.open({
         size: 'small',
-        description: this.labels.feedback,
-        options: this.options,
+        liked: false,
+        label: this.labels.feedbackHelpUsImprove,
         handleSubmit: this.submitFeedback.bind(this),
-        optionsLabel: this.labels.whyGeneratedAnswerWasNotHelpful,
       });
```
We could isolate the logic to open the modal in a separate function so we can just call it for the like handler and the dislike handler instead of having to write this same logic twice.

erocheleau replied:
I don't see much value in doing this, especially when we have at least 1 difference

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/__tests__/quanticFeedbackModalQna.test.js:
```diff
@@ -0,0 +1,299 @@
+  // All base components in Jest tests are stubs. So we have to set the value manually.
+  // Including to the checkValidity method.
+  lightningElement.value = value;
+  lightningElement.validity = true;
```
is it necessary to set `validity` to true? forcing `checkValidity` to return `true` should be enough I think

erocheleau replied:
🤷 not a huge issue to have both huh?

mmitiche replied:
just so to slightly reduce the complexity, if we don't need it, why adding it.

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/__tests__/quanticFeedbackModalQna.test.js:
```diff
@@ -0,0 +1,299 @@
+function lightningElementValueChange(element, selector, value) {
+  const lightningElement = element.shadowRoot.querySelector(selector);
+  // All base components in Jest tests are stubs. So we have to set the value manually.
+  // Including to the checkValidity method.
```
I think this comment should be adjusted,

erocheleau replied:
To what?

mmitiche replied:
It's just that `Including to the checkValidity method` doesn't seem like a correct english sentence x)

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/__tests__/quanticFeedbackModalQna.test.js:
```diff
@@ -0,0 +1,299 @@
+        lightningElementValueChange(
+          element,
+          selectors.questionHallucinationRadio,
+          'yes'
```
I wouldn't put all the expected answers to `yes` to also check that our component can return other values as well and not always `yes`  whatever what option you select.

erocheleau replied:
It doesn't matter which value I put, I'm the one setting the value.

So even if I put 'abracadabra', in the end I'm the one that says `value='abracadabra'` and then executing `get value` so essentially I can only test that the code gets the value I set. Changing the string doesn't add anything to the test.

Keep in mind, in reality I am not clicking, nor selecting an option, I am setting the `lightning-radio-group` stub's `value` attribute to any string I want. And when the code of `quanticFeedbackModalQna` checks for the value, it simply returns the value I've set on the stub.

mmitiche replied:
That's where we feel the value of E2E tests to come and complement the unit tests.

mmitiche reviewed (commented):
Very nice job! just few adjustments suggested👍

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/quanticGeneratedAnswer.js:
```diff
@@ -317,11 +335,12 @@ export default class QuanticGeneratedAnswer extends LightningElement {
+    //   this.generatedAnswer.sendDetailedFeedback(feedbackPayload.details);
+    // } else if (feedbackPayload?.value) {
+    //   this.generatedAnswer.sendFeedback(feedbackPayload.value);
+    // }
```
huh? that was gone for a while when you reviewed this no?

mmitiche replied:
I know that we skipped the tests, but the logic was not removed before this pr

mmitiche replied:
Ah I see I see x) all good then x)

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/quanticFeedbackModalQna.css:
```diff
@@ -0,0 +1,7 @@
+:host {
+    --slds-c-textarea-sizing-min-height: 100px;
+}
+
+.feedback-modal-qna__radio-group {
+    min-width: 186px;
```
yea, I can calculate it

Not sure what I want to happen if we increase the font size by a lot though

erocheleau replied:
This still feels like black magic :D but it's better now.
And this was also the cause of the alignment being weird on small screens.

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/quanticFeedbackModalQna.js:
```diff
@@ -0,0 +1,188 @@
+// @ts-nocheck
```
Damn I forgot I put that, it is in the quanticFeedbackModal.js file as well xD

I'll see if I can fix it.

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/quanticFeedbackModalQna.js:
```diff
@@ -0,0 +1,188 @@
+ * The `QuanticFeedbackModalQna` component overlays a message modal on top of the current app window, the modal contains a form that allows the user to give feedback on a CRGA response.
+ *
+ * Under the hood, the component relies on a [`lightningModal`](https://developer.salesforce.com/docs/component-library/bundle/lightning-modal/documentation) component.
+ * For an example of how to use the `QuanticFeedbackModal` component, see the [`quanticSmartSnippet`](https://github.com/coveo/ui-kit/blob/master/packages/quantic/force-app/main/default/lwc/quanticSmartSnippet/quanticSmartSnippet.js) implementation.
+ *
```
Updated to what?

mmitiche replied:
we need to talk about the `QuanticFeedbackModalQna` instead of talking about `QuanticFeedbackModal`.

erocheleau replied:
ohhh yea

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/quanticFeedbackModalQna.js:
```diff
@@ -0,0 +1,188 @@
+    return this.template.querySelectorAll('lightning-radio-group');
+  }
+
+  get feedbackFormIsValid() {
```
Yea that's true I'll change it

erocheleau replied:
the `feedbackFormIsValue` is an actual property of the feedback form.

I could change it to a function as well and it would be perfectly fine to me as well.

That one was taken from an example in the salesforce doc, so it's more a property than a function for me.

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/quanticFeedbackModalQna.js:
```diff
@@ -0,0 +1,188 @@
+  }
+
+  documentUrlChangeHandler(event) {
+    this.documentUrlValue = event?.target?.value;
```
I guess not

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticFeedbackModalQna/quanticFeedbackModalQna.js:
```diff
@@ -0,0 +1,188 @@
+    },
+    {
+      question: this.labels.questionHallucination,
+      changeFn: (event) => {
+        this.hallucinationFreeValue = event.target.value;
+      },
```
Sure I can try that.

erocheleau replied:
I tried and it doesn't fix it.
I can reproduce the same thing with the default `lightning-radio-group` even in the Salesforce lightning component examples, so I think it's on them...

mmitiche reviewed (approved):
Well done 👍

erocheleau commented:
Following @lvu285 review:

1. I changed it so that if you don't submit your feedback (click on vote button, but click skip or X), if you click on the same vote button, it re-opens the modal.
2. Even if you submit your feedback, you can change your vote, but cannot re-open the modal, even if you change your vote.
3. Also changed the look on really smaller screens:

![image](https://github.com/user-attachments/assets/29292504-9bf4-4eb7-a58c-4cfddc319860)

mmitiche reviewed (approved):
latest changes look good to me 👍
